### DISCUSSION TITLE
Cannot seem to get more than 3 NIO threads despite configuration

### DISCUSSION THREAD

**Discussion by drub0y** [redis/lettuce#2923] [Q&A]
Setup:
 - `3` node Redis cluster
 - Using Reactive APIs _through Spring Redis_
 - Lettuce configured with
   - Fixed `ioThreadPoolSize` of `8`
   - Fixed `computationThreadPoolSize` of `32`
   - Set `publishOnScheduler` to true to get all deserialization off NIO threads
   - connection pool with min of `3` and max of `12`

Spring configuration:

```Java
@Bean
 public LettuceClientConfigurationBuilderCustomizer lettuceClientConfigurationBuilderCustomizer() {
   return clientConfigurationBuilder -> clientConfigurationBuilder
      .clientResources(DefaultClientResources.builder()
          .ioThreadPoolSize(8)
          .computationThreadPoolSize(32)
          .build())
      .clientOptions(ClientOptions.builder()
          .publishOnScheduler(true)
          .build());
    }

```

I have confirmed that I see this logic running and can confirm, through the debugger, that it is telling the client it should use a thread pool of `8` `NioEventLoop` instances and this seems to bear out:

![image](https://github.com/user-attachments/assets/92fb1a24-9eb9-443b-adf0-3d179fd7b592)

I can also see that it has clearly adopted my connection pooling policy of `3` min, `12` max, etc:

![image](https://github.com/user-attachments/assets/07761069-fa29-448d-8574-6e63e3a53ec6)
 
Despite the configuration above, I only ever see `3` Lettuce NIO threads being spun up and used under our load tests:

![image](https://github.com/user-attachments/assets/7de0e0b4-dab1-4f7a-b484-a76e2a51c544)

This is a 1min slice under heavy load and here's a detailed breakdown of what the threads are doing during that time:

![image](https://github.com/user-attachments/assets/5c0a00be-6e13-456a-a721-5e9231f6a844)

> Side note: The setting of `32` computation threads works exactly as intended and I see whatever number I configure impacting that thread pool the way I expect.

We are hitting bottleneck pretty quickly where the command queues are backing up because the NIO threads are just maxed out trying to keep up with the volume of commands. Meanwhile our Redis nodes are only at like 15% and the rest of the pod's resources are fine.

The documentation is not very clear on this TBH, but my understanding at the moment is that, to achieve max throughput, we want a combination of both more connections as well as more NIO threads. This is because connections can only use one NIO thread at a time, but a NIO thread can serve multiple connections over time so it need not be 1:1. I've, of course, consulted our AI overlords, but they only seem to confirm my understanding is accurate. That said, clearly there are `3` Redis nodes and I'm ending up with `3` NIO threads only ever in use, so perhaps I'm missing something about the math or just not interpreting the docs correctly. 

Finally, to head off any questions, there was a point where some of the code was using `executeInSession` which I understand would force some sort of sequential, connection bound execution, but all of the code at this point is _read-only_ commands executed in parallel _at all times_ using proper reactive patterns. If it weren't for the limitation and subsequent saturation of the `3` NIO threads everything else seems perfect.

Any advice welcome and if I can provide any more detail please lmk!

**Comment by drub0y** [comment]
As I go back and read through to see what I might be misunderstanding, [this was a paragraph that had me originally remove connection pooling altogether](https://github.com/redis/lettuce/wiki/Connection-Pooling#is-connection-pooling-necessary). I assume that, since we're fully async, we're ultimately better off _not_ using connection pooling at all... especially since it does not seem to have improved things.

**Reply by tishun** [reply]
Yes, connection pooling is only advised in scenarios where you need transactions across different threads or some of the threads use blocking commands. For most cases connection polling will reduce the performance of the driver, as it mainly relies on multiplexing the threads over a single connection.

**Comment by drub0y** [comment]
Also, some feedback on the docs, but from [the docs on client configuration](https://github.com/redis/lettuce/wiki/Configuring-Client-resources#configuration-settings), this particular paragraph on `ioThreadPoolSize` doesn't quite make sense to me personally:

> The number of threads in the I/O thread pools. The number defaults to the number of available processors that the runtime returns (which, as a well-known fact, sometimes does not represent the actual number of processors). Every thread represents an internal event loop where all I/O tasks are run. The number does not reflect the actual number of I/O threads because the client requires different thread pools for Network (NIO) and Unix Domain Socket (EPoll) connections. The minimum I/O threads are 3. A pool with fewer threads can cause undefined behavior.

Specifically, what does this mean?

> The number does not reflect the actual number of I/O threads because the client requires different thread pools for Network (NIO) and Unix Domain Socket (EPoll) connections.

If it doesn't reflect the actual number of threads, what does the number I'm setting actually _mean_ and how should I see it reflected if not by some threads? 🤔 

Clearly, I see _some_ `lettuce-nioEventLoop-#` threads in my service and they are _very_ busy, so I would expect increasing this number would result in more of them so they can spread some load around to other available CPU cores.

**Comment by drub0y** [comment]
Well, I might have found the answer [here](https://github.com/redis/lettuce/wiki/Redis-Cluster#connection-count-for-a-redis-cluster-connection-object):

> Connections are allocated on demand and not up-front to start with a minimal set of connections. 
> Formula to calculate the maximum number of transport connections for a single connection object:
>
> 1 + (N * 2)
> Where N is the number of cluster nodes.

So, if I'm understanding correctly, because we're talking to a Redis cluster the connection patterns completely change and none of the other stuff I've mentioned matters because this is how cluster connections work where only `N` (where `N` == number of nodes) transport connections will actually ever be used.

Can anyone from the team confirm this?

**Reply by tishun** [reply]
I can confirm that - in case of a cluster connection - the formula above should be correct.

Note that "transport connections" only includes connections that are used to send commands to Redis, the toal number of all connections might (and probably is) higher due to topology refresh and etc. connections

**Comment by drub0y** [comment]
FWIW, I'm convinced that the conclusions I've drawn in this thread are accurate and it's backed up by what I'm actually seeing at runtime. Hopefully this helps some others who might be confused, but going to close the question out.

**Comment by tishun** [comment]
I am still a bit confused about what the issue is.

> We are hitting bottleneck pretty quickly where the command queues are backing up because the NIO threads are just maxed out trying to keep up with the volume of commands. Meanwhile our Redis nodes are only at like 15% and the rest of the pod's resources are fine.

Not sure what "commands queues are backing up" and "NIO threads are just maxed out" means. What is the actual symptom you are seeing, is it memory footprint, CPU utilization, errors being thrown? 

>  ... to achieve max throughput, we want a combination of both more connections as well as more NIO threads

It really depends on what the problem is. You might be underutilizing your CPU (theoretically), or you might be hitting the bandwidth limits of the network ... to be of help I need to know more about your setup and the exact issue you are seeing.
